x/subscription/keeper/grpc_query_estimated_rewards.go

```diff
@@ -0,0 +1,103 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	epochstoragetypes "github.com/lavanet/lava/v2/x/epochstorage/types"
+	rewardstypes "github.com/lavanet/lava/v2/x/rewards/types"
+	"github.com/lavanet/lava/v2/x/subscription/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (k Keeper) EstimatedRewards(goCtx context.Context, req *types.QueryEstimatedRewardsRequest) (*types.QueryEstimatedRewardsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	res := types.QueryEstimatedRewardsResponse{}
+
+	delegationAmount, err := sdk.ParseCoinNormalized(req.Amount)
+	if err != nil {
+		return nil, err
+	}
+
+	storage := k.epochstorageKeeper.GetAllStakeEntriesCurrentForChainId(ctx, req.ChainId)
+
+	totalStake := math.ZeroInt()
+	var entry epochstoragetypes.StakeEntry
+	found := false
+	for _, e := range storage {
+		totalStake = totalStake.Add(e.EffectiveStake())
+		if e.Address == req.Provider {
+			found = true
+			entry = e
+		}
+	}
+	if !found {
+		return nil, fmt.Errorf("provider not found in stake entries for chain ID: %s", req.ChainId)
+	}
+
+	entry.DelegateTotal = entry.DelegateTotal.Add(delegationAmount)
+	delegatorPart := sdk.NewDecFromInt(delegationAmount.Amount).QuoInt(totalStake)
+	if entry.DelegateLimit.Amount.LT(entry.DelegateTotal.Amount) {
+		delegatorPart = delegatorPart.MulInt(entry.DelegateLimit.Amount).QuoInt(entry.DelegateTotal.Amount)
+	}
+
+	totalSubsRewards := sdk.Coins{}
+	subsIndices := k.GetAllSubscriptionsIndices(ctx)
+	for _, subIndex := range subsIndices {
+		sub, found := k.GetSubscription(ctx, subIndex)
+		if found {
+			sub.Credit.Amount = sub.Credit.Amount.Quo(sdk.NewIntFromUint64(sub.DurationLeft))
+			totalSubsRewards = totalSubsRewards.Add(sub.Credit)
+		}
+	}
+
+	emisions := k.rewardsKeeper.SpecEmissionParts(ctx)
+	var specEmission rewardstypes.SpecEmissionPart
+	found = false
+	for _, k := range emisions {
+		if k.ChainID == req.ChainId {
+			specEmission = k
+			found = true
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("spec emission part not found for chain ID: %s", req.ChainId)
+	}
+
+	subscriptionRewards := sdk.NewCoins(totalSubsRewards.MulInt(specEmission.Emission.MulInt64(100).RoundInt())...)
```

Comment:
a valid assumption but might not always be true so add this assumption to the query help

Reply:
done